qiime2/metadata/metadata.py

```diff
@@ -353,7 +324,49 @@ def column_count(self):
         """
         return len(self._columns)
 
-    def __init__(self, dataframe):
+    @property
+    def artifacts(self):
+        """Artifacts that are the source of the metadata.
+
+        This property is read-only.
+
+        Returns
+        -------
+        tuple of qiime2.Artifact
+            Source artifacts of the metadata.
+
+        """
+        if self._source_artifact is not None:
+            return (self._source_artifact,)
+
+        artifacts = set()
+        for source in self._column_sources.values():
+            if source is not None:
+                artifacts.update(artifact for artifact in source.artifacts)
+        return tuple(artifacts)
+
+    @property
+    def source_artifact(self):
+        """Artifact that is the source of the metadata.
+
+        This property is read-only.
+
+        Returns
+        -------
+        qiime2.Artifact
+            Source artifact of the metadata.
+
+        """
+        return self.source_artifact
+
+    # Try to ensure the source artifact is actually an artifact
+    def _add_source_artifact(self, artifact):
+        if not isinstance(artifact, qiime2.Artifact):
+            raise TypeError('Source Artifact must a an Artifact object not '
+                            f'{artifact}.')
+        self._source_artifact = artifact
+
+    def __init__(self, dataframe, column_names=None, column_sources=None):
```

Comment:
Should we use a private constructor for handling `column_names` & `column_sources`, instead? I think the current signature will be confusing to Artifact API users.